src/Microsoft.AspNetCore.Http.Abstractions/Routing/EndpointMetadataCollection.cs

```diff
@@ -0,0 +1,178 @@
+        /// </summary>
+        /// <param name="index">The index of the item to retrieve.</param>
+        /// <returns>The item at <paramref name="index"/>.</returns>
+        public object this[int index] => _items[index];
+
+        /// <summary>
+        /// Gets the count of metadata items.
+        /// </summary>
+        public int Count => _items.Length;
+
+        /// <summary>
+        /// Gets the most significant metadata item of type <typeparamref name="T"/>.
+        /// </summary>
+        /// <typeparam name="T">The type of metadata to retrieve.</typeparam>
+        /// <returns>
+        /// The most significant metadata of type <typeparamref name="T"/> or <c>null</c>.
+        /// </returns>
+        public T GetMetadata<T>() where T : class
+        {
+            for (var i = _items.Length - 1; i >= 0; i--)
+            {
+                var item = _items[i] as T;
+                if (item != null)
+                {
+                    return item;
+                }
+            }
+
+            return default;
+        }
+
+        /// <summary>
+        /// Gets the metadata items of type <typeparamref name="T"/> in ascending
+        /// order of precedence.
+        /// </summary>
+        /// <typeparam name="T">The type of metadata.</typeparam>
+        /// <returns>A sequence of metadata items of <typeparamref name="T"/>.</returns>
+        public IEnumerable<T> GetOrderedMetadata<T>() where T : class
+        {
+            for (var i = 0; i < _items.Length; i++)
+            {
+                var item = _items[i] as T;
+                if (item != null)
+                {
+                    yield return item;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Gets an <see cref="IEnumerator"/> of all metadata items.
+        /// </summary>
+        /// <returns>An <see cref="IEnumerator"/> of all metadata items.</returns>
+        public Enumerator GetEnumerator() => new Enumerator(this);
+
+        /// <summary>
+        /// Gets an <see cref="IEnumerator{Object}"/> of all metadata items.
+        /// </summary>
+        /// <returns>An <see cref="IEnumerator{Object}"/> of all metadata items.</returns>
+        IEnumerator<object> IEnumerable<object>.GetEnumerator() => GetEnumerator();
+
+        /// <summary>
+        /// Gets an <see cref="IEnumerator"/> of all metadata items.
+        /// </summary>
+        /// <returns>An <see cref="IEnumerator"/> of all metadata items.</returns>
+        IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
+
+        /// <summary>
+        /// Enumerates the elements of an <see cref="EndpointMetadataCollection"/>.
+        /// </summary>
+        public struct Enumerator : IEnumerator<object>
+        {
+            // Intentionally not readonly to prevent defensive struct copies
+            private object[] _items;
+            private int _index;
+            private object _current;
```

Comment:
You could get away with not having this field (just index in to the array as part of `Current`)

Reply:
Yeah, please don't change this, it was done this way on purpose